crypto/bindings/srs.ts

```diff
@@ -0,0 +1,215 @@
+  type Cache,
+  type CacheHeader,
+} from '../../../lib/proof-system/cache.js';
+import { assert } from '../../../lib/errors.js';
+import { MlArray, MlOption } from '../../../lib/ml/base.js';
+import { OrInfinity, OrInfinityJson } from './curve.js';
+
+export { srs, setSrsCache, unsetSrsCache };
+
+type WasmSrs = WasmFpSrs | WasmFqSrs;
+
+type SrsStore = Record<number, WasmSrs>;
+
+function empty(): SrsStore {
+  return {};
+}
+
+const srsStore = { fp: empty(), fq: empty() };
+
+let cache: Cache | undefined;
+
+function setSrsCache(c: Cache) {
+  cache = c;
+}
+function unsetSrsCache() {
+  cache = undefined;
+}
+
+const srsVersion = 1;
+
+function cacheHeaderLagrange(f: 'fp' | 'fq', domainSize: number): CacheHeader {
+  let id = `lagrange-basis-${f}-${domainSize}`;
+  return withVersion(
+    {
+      kind: 'lagrange-basis',
+      persistentId: id,
+      uniqueId: id,
+      dataType: 'string',
+    },
+    srsVersion
+  );
+}
+function cacheHeaderSrs(f: 'fp' | 'fq', domainSize: number): CacheHeader {
+  let id = `srs-${f}-${domainSize}`;
+  return withVersion(
+    {
+      kind: 'srs',
+      persistentId: id,
+      uniqueId: id,
+      dataType: 'string',
+    },
+    srsVersion
+  );
+}
+
+function srs(wasm: Wasm, conversion: RustConversion) {
+  return {
+    fp: srsPerField('fp', wasm, conversion),
+    fq: srsPerField('fq', wasm, conversion),
+  };
+}
+
+function srsPerField(f: 'fp' | 'fq', wasm: Wasm, conversion: RustConversion) {
+  // note: these functions are properly typed, thanks to TS template literal types
+  let createSrs = wasm[`caml_${f}_srs_create_parallel`];
+  let getSrs = wasm[`caml_${f}_srs_get`];
+  let setSrs = wasm[`caml_${f}_srs_set`];
+
+  let maybeLagrangeCommitment = wasm[`caml_${f}_srs_maybe_lagrange_commitment`];
+  let lagrangeCommitment = wasm[`caml_${f}_srs_lagrange_commitment`];
+  let setLagrangeBasis = wasm[`caml_${f}_srs_set_lagrange_basis`];
+  let getLagrangeBasis = wasm[`caml_${f}_srs_get_lagrange_basis`];
+
+  return {
+    /**
+     * returns existing stored SRS or falls back to creating a new one
+     */
+    create(size: number): WasmSrs {
+      let srs = srsStore[f][size] as WasmSrs | undefined;
+
+      if (srs === undefined) {
+        if (cache === undefined) {
+          // if there is no cache, create SRS in memory
+          srs = createSrs(size);
+        } else {
+          let header = cacheHeaderSrs(f, size);
+
+          // try to read SRS from cache / recompute and write if not found
+          try {
+            let bytes = cache.read(header);
+
+            // TODO: this takes a bit to long, about 300ms for 2^16
+            // `pointsToRust` is the clear bottleneck
+            let jsonSrs: OrInfinityJson[] = JSON.parse(
+              new TextDecoder().decode(bytes)
+            );
+            let mlSrs = MlArray.mapTo(jsonSrs, OrInfinity.fromJSON);
+            let wasmSrs = conversion[f].pointsToRust(mlSrs);
+
+            srs = setSrs(wasmSrs);
+          } catch (e) {
```

Comment:
should it not be around cache.read then? We must be sure it comes from the cache.read function. Many other exceptions can be raised.

Reply:
hm, but `read()` can be very simple. it just returns bytes contained in a file. what if the file is garbage? then the decoding here will fail, not `read()`. I think we shouldn't trust the file system to return something that can be deserialized. but we also shouldn't treat a failed cache read as a hard error that crashes the process, since we can just regenerate the data and write back a correct version.

maybe to address your point, the conversion logic should be tested as a unit. and there should be an easy way to test a `Cache`. maybe we can just introduce a "caches may fail" mode which is used when debugging your cache, where errors aren't swallowed

Reply:
Something explicit in case of an error would be nice to have. It will help debugging in this case, even not very often happening.

Reply:
I'm addressing this by adding a `debug` parameter to the `Cache`. Also I'm abstracting away the try-catch acrobatics into helpers